substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/MemoryUtil.java

```diff
@@ -82,23 +93,40 @@ public static List<String> determineMemoryFlags(NativeImage.HostFlags hostFlags)
     }
 
     /**
-     * Returns a percentage (0.0-100.0) to be used as a value for the -XX:MaxRAMPercentage flag of
-     * the builder process. Prefer free memory over total memory to reduce memory pressure on the
-     * host machine. Note that this method uses OperatingSystemMXBean, which is container-aware.
+     * Returns a percentage (0.0-100.0) to be used as a value for the -XX:MaxRAMPercentage or
+     * -XX:MaximumHeapSizePercent flags of the builder process. Dedicated mode uses a fixed
+     * percentage of total memory and is the default in containers. Shared mode tries to use
+     * available memory to reduce memory pressure on the host machine. Note that this method uses
+     * OperatingSystemMXBean, which is container-aware.
      */
-    private static double determineReasonableMaxRAMPercentage() {
+    private static List<String> determineReasonableMaxRAMPercentage(Function<Double, String> toMemoryFlag) {
         var osBean = (com.sun.management.OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean();
         final double totalMemorySize = osBean.getTotalMemorySize();
-        double reasonableMaxMemorySize = osBean.getFreeMemorySize();
+        final double dedicatedMemorySize = totalMemorySize * DEDICATED_MODE_TOTAL_MEMORY_RATIO;
 
-        if (reasonableMaxMemorySize < DEDICATED_MODE_THRESHOLD) {
-            /*
-             * When free memory is low, for example in memory-constrained environments or when a
-             * good amount of memory is used for caching, use a fixed percentage of total memory
-             * rather than free memory. In containerized environments, builds are expected to run
-             * more or less exclusively (builder + driver + optional Gradle/Maven process).
-             */
-            reasonableMaxMemorySize = totalMemorySize * DEDICATED_MODE_TOTAL_MEMORY_RATIO;
+        String memoryUsageReason = "unknown";
+        final boolean isDedicatedMemoryUsage;
+        if ("true".equals(System.getenv("CI"))) {
+            isDedicatedMemoryUsage = true;
+            memoryUsageReason = "$CI set";
+        } else if (isContainerized()) {
+            isDedicatedMemoryUsage = true;
+            memoryUsageReason = "in container";
+        } else {
+            isDedicatedMemoryUsage = false;
+        }
+
+        double reasonableMaxMemorySize;
+        if (isDedicatedMemoryUsage) {
+            reasonableMaxMemorySize = dedicatedMemorySize;
+        } else {
+            reasonableMaxMemorySize = getAvailableMemorySize();
+            if (reasonableMaxMemorySize >= MIN_AVAILABLE_MEMORY_THRESHOLD) {
+                memoryUsageReason = "enough available";
+            } else { // fall back to dedicated mode
+                memoryUsageReason = "not enough available";
```

Comment:
"(not) enough available" seems confusing to me.

* "enough available" makes me think I should not tune anything even if my build fails due to out of memory or I get a lot of time spent on GC. I would replace this with something like "available free memory used".
* "not enough available" makes me feel something is wrong and I need to tune things.  I would replace this with something like "less than 8GB ram is available, setting mac to 85% of system memory".

Reply:
Yes, we can do that. I also tried to keep the text short so that the lines fit in 80 characters. I'll see if I can come up with something better.

Reply:
Changed it to:
```
 - 13.24GB of memory (85.0% of system memory, less than 8GB of memory available)
```
for now.